src/opm/output/data/InterRegFlowMap.cpp

```diff
@@ -0,0 +1,586 @@
+        std::swap(low, high);
+    }
+
+    this->i_.push_back(low);
+    this->j_.push_back(high);
+
+    this->max_i_ = std::max(this->max_i_, this->i_.back());
+    this->max_j_ = std::max(this->max_j_, this->j_.back());
+
+    const auto start = this->v_.size();
+    this->v_.insert(this->v_.end(), Window::bufferSize(), ElmT{0});
+
+    Window{ this->v_.begin() + start, this->v_.end() }.addFlow(sign, v);
+}
+
+void
+Opm::data::InterRegFlowMap::
+Connections::add(const int         maxRowIdx,
+                 const int         maxColIdx,
+                 const Neighbours& rows,
+                 const Neighbours& cols,
+                 const RateBuffer& rates)
+{
+    this->i_.insert(this->i_.end(), rows .begin(), rows .end());
+    this->j_.insert(this->j_.end(), cols .begin(), cols .end());
+    this->v_.insert(this->v_.end(), rates.begin(), rates.end());
+
+    this->max_i_ = std::max(this->max_i_, maxRowIdx);
+    this->max_j_ = std::max(this->max_j_, maxColIdx);
+}
+
+void Opm::data::InterRegFlowMap::Connections::clear()
+{
+    this->v_.clear();
+    this->j_.clear();
+    this->i_.clear();
+
+    this->max_i_ = -1;
+    this->max_j_ = -1;
+}
+
+bool Opm::data::InterRegFlowMap::Connections::empty() const
+{
+    return this->i_.empty();
+}
+
+bool Opm::data::InterRegFlowMap::Connections::isValid() const
+{
+    return (this->i_.size() == this->j_.size())
+        && (this->v_.size() == this->i_.size()*Window::bufferSize());
+}
+
+int Opm::data::InterRegFlowMap::Connections::maxRow() const
+{
+    return this->max_i_;
+}
+
+int Opm::data::InterRegFlowMap::Connections::maxCol() const
+{
+    return this->max_j_;
+}
+
+Opm::data::InterRegFlowMap::Neighbours::size_type
+Opm::data::InterRegFlowMap::Connections::nnz() const
+{
+    return this->i_.size();
+}
+
+const Opm::data::InterRegFlowMap::Neighbours&
+Opm::data::InterRegFlowMap::Connections::i() const
+{
+    return this->i_;
+}
+
+const Opm::data::InterRegFlowMap::Neighbours&
+Opm::data::InterRegFlowMap::Connections::j() const
+{
+    return this->j_;
+}
+
+const Opm::data::InterRegFlowMap::RateBuffer&
+Opm::data::InterRegFlowMap::Connections::v() const
+{
+    return this->v_;
+}
+
+// =====================================================================
+
+// ---------------------------------------------------------------------
+// Class Opm::data::InterRegFlowMap::CSR
+// ---------------------------------------------------------------------
+
+void
+Opm::data::InterRegFlowMap::
+CSR::merge(const Connections& conns, const Offset numRegions)
+{
+    if (! conns.empty() &&
+        (static_cast<Offset>(conns.maxRow()) >= numRegions))
+    {
+        throw std::invalid_argument {
+            "Input graph contains more "
+            "source regions than are "
+            "implied by explicit size of "
+            "adjacency matrix"
+        };
+    }
+
+    this->assemble(conns.i(), conns.j(), conns.maxRow(), conns.maxCol());
+    this->compress(numRegions, conns.v());
+}
```

Comment:
There you go. I am definitely not smart enough to understand this. Please bear with me.

Looking at the other PR Connections can look like this: `{{r1, r2, rate1},{r1, r3, rate2}, {r1, r2, rate3, ....}`. There seems to be the need to accumulate the rates for connections that appear multiple times, e.g. by sorting for the region pairs and then adding those with the same values up. I cannot find this and it cannot be done in assemble or compress (because it would need i, j and v). Where is this done?

Reply:
> There seems to be the need to accumulate the rates for connections that appear multiple times, e.g. by sorting for the region pairs and then adding those with the same values up. I cannot find this and it cannot be done in assemble or compress (because it would need i, j and v). Where is this done?

The direct summation of duplicate elements happens in the [`CSR::accumulateFlowRates()`](https://github.com/OPM/opm-common/blob/7c2f1a10b7e12e62c4cef2574c5bb82e1e2934d2/src/opm/output/data/InterRegFlowMap.cpp#L378-L381) member function.  `AccumulateFlowRates()` is the [last stage](https://github.com/OPM/opm-common/blob/7c2f1a10b7e12e62c4cef2574c5bb82e1e2934d2/src/opm/output/data/InterRegFlowMap.cpp#L296-L301) of `compress()`.  This implementation uses an ancillary mapping array ([`CSR::compressedIdx_`](https://github.com/OPM/opm-common/blob/7c2f1a10b7e12e62c4cef2574c5bb82e1e2934d2/opm/output/data/InterRegFlowMap.hpp#L225-L226), named `elemIdx_` in the first version of this PR but I renamed it to hopefully make its purpose more clear) that tracks which unique element in the compressed array each possibly duplicate uncompressed element contributes to.  I suppose you could argue that this strategy is too clever/unusual, but I'll submit that it also has a couple of really nice properties&mdash;not least of which is that this mapping array enables decoupling the formation of the graph structure from actually accumulating duplicate entries.  It will also enable O(1) rather than O(log k) subsequent assembly per element if we guarantee that we always visit the connections in the same order.

In any case, we form the initial `compressedIdx_` mapping as part of [creating](https://github.com/OPM/opm-common/blob/7c2f1a10b7e12e62c4cef2574c5bb82e1e2934d2/src/opm/output/data/InterRegFlowMap.cpp#L430-L436) the initial, unsorted pair of row pointer ("IA") and column index ("JA") arrays.  Then, as part of [sorting](https://github.com/OPM/opm-common/blob/7c2f1a10b7e12e62c4cef2574c5bb82e1e2934d2/src/opm/output/data/InterRegFlowMap.cpp#L311-L320) the column indices we take care to also maintain the [`compressedIdx_` mapping](https://github.com/OPM/opm-common/blob/7c2f1a10b7e12e62c4cef2574c5bb82e1e2934d2/src/opm/output/data/InterRegFlowMap.cpp#L454).  The final step is to account for [removing duplicate elements](https://github.com/OPM/opm-common/blob/7c2f1a10b7e12e62c4cef2574c5bb82e1e2934d2/src/opm/output/data/InterRegFlowMap.cpp#L476-L480) as part of [`CSR::condenseDuplicates()`](https://github.com/OPM/opm-common/blob/7c2f1a10b7e12e62c4cef2574c5bb82e1e2934d2/src/opm/output/data/InterRegFlowMap.cpp#L341).